src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolBlockingQueue.AnyOS.cs

```diff
@@ -0,0 +1,219 @@
+        }
+
+        private void MarkCompleted()
+        {
+            int count = Volatile.Read(ref _processingThreads);
+
+            Debug.Assert(count > 0);
+            do
+            {
+                int prevCount = Interlocked.CompareExchange(ref _processingThreads, count - 1, count);
+                if (prevCount == count)
+                {
+                    break;
+                }
+                count = prevCount;
+            } while (count > 0);
+        }
+
+        public static bool RequiresMitigation(object? workItem)
+        {
+            object? entryPoint = GetEntryPoint(workItem);
+            if (entryPoint == null)
+            {
+                return false;
+            }
+
+            return s_blockingItems?.Contains(entryPoint) ?? false;
+        }
+
+        public static void RegisterForBlockingDetection(object? workItem)
+            => t_currentWorkItem = workItem;
+
+        public static void ClearRegistration()
+            => t_currentWorkItem = null;
+
+        public static void Enable()
+        {
+            if (!s_enabled)
+            {
+                Volatile.Write(ref s_enabled, true);
+            }
+
+            object? workItem = t_currentWorkItem;
+            if (workItem is null) return;
+
+            object? entryPoint = GetEntryPoint(workItem);
+            if (entryPoint == null)
+            {
+                return;
+            }
+
+            HashSet<object>? items = s_blockingItems;
+            if (items?.Contains(entryPoint) ?? false)
+            {
+                // Already there
+                return;
+            }
+
+            // Lock as may be a batch of same items and we don't want to
+            // create a large amount of HashSet copies for the same item.
+            lock (Processor)
+            {
+                if (items?.Contains(entryPoint) ?? false)
+                {
+                    // Already there
+                    return;
+                }
+
+                // Copy on add; so can be used without locks when checking for existance
+                items = items is null ? new HashSet<object>() : new HashSet<object>(items);
+                items.Add(entryPoint);
+
+                s_blockingItems = items;
+            }
+        }
+
+        private static object? GetEntryPoint(object? workItem)
+        {
+            if (workItem is null || ReferenceEquals(workItem, s_processor))
+            {
+                // Don't mark self
+                return null;
+            }
+
+            if (workItem is QueueUserWorkItemCallbackBase quwi)
+            {
+                return quwi._delegate.GetMethodImpl();
+            }
+            if (workItem is IAsyncStateMachineBox sm)
+            {
+                return sm.GetType();
+            }
+            if (workItem is Task t)
+            {
+                return t.m_action?.GetMethodImpl();
+            }
+
+            return workItem?.GetType();
```

Comment:
```suggestion
            return workItem switch
            {
                QueueUserWorkItemCallbackBase quwi => quwi._delegate.GetMethodImpl(),
                 IAsyncStateMachineBox sm => sm.GetType(),
                 Task t => t.m_action?.GetMethodImpl(),
                 _ =>  workItem?.GetType()
            };
```

Does this work as a switch expression instead?